rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        write(b1_off, src);
+    } else {
+        // Case: Staged reads, one per block that holds requested data.
+
+        let mut i = 0; // invariant: i = # of bytes stored.
+        let mut s = src.as_ptr(); // source for bytes.
+        let mut d = b1_off; // dest of bytes, as absolute index.o
+
+        // do initial write (a special case, generally not full block length).
+        write(d, core::slice::from_raw_parts(s, b1_len as usize));
+
+        // Advance input and output positions (i, s and d respectively).
+        i += b1_len;
+        s = s.offset(b1_len as isize);
+
+        // Do rest of block-sized reads.
+        // (invariant: they always occur at the start of a block).
+        loop {
+            let (d_, _, b_len) = r.relative_into_absolute_info(offset + i);
+            d = d_;
+            if i + b_len > src.len() as u64 {
+                // case: last (generally partial) block.
+                if src.len() as u64 > i {
+                    write(d, core::slice::from_raw_parts(s, src.len() - i as usize));
+                }
+                break;
+            } else {
+                // case: internal (full) block.
+                write(d, core::slice::from_raw_parts(s, BLOCK_IN_BYTES as usize));
+                s = s.offset(BLOCK_IN_BYTES as isize);
+                i += BLOCK_IN_BYTES;
+            }
+        }
+    }
+}
+
+// -- Region load operations.
+
+#[ic_mem_fn]
+pub unsafe fn region_load_word8<M: Memory>(mem: &mut M, r: Value, offset: u64) -> u32 {
+    let mut byte: [u8; 1] = [0];
+    region_load(mem, r, offset, &mut byte);
+    byte[0] as u32
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_word16<M: Memory>(mem: &mut M, r: Value, offset: u64) -> u32 {
+    let mut bytes: [u8; 2] = [0; 2];
+    region_load(mem, r, offset, &mut bytes);
+    core::primitive::u16::from_le_bytes(bytes).into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_word32<M: Memory>(mem: &mut M, r: Value, offset: u64) -> u32 {
+    let mut bytes: [u8; 4] = [0; 4];
+    region_load(mem, r, offset, &mut bytes);
+    core::primitive::u32::from_le_bytes(bytes).into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_word64<M: Memory>(mem: &mut M, r: Value, offset: u64) -> u64 {
+    let mut bytes: [u8; 8] = [0; 8];
+    region_load(mem, r, offset, &mut bytes);
+    core::primitive::u64::from_le_bytes(bytes).into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_float64<M: Memory>(mem: &mut M, r: Value, offset: u64) -> f64 {
+    let mut bytes: [u8; 8] = [0; 8];
+    region_load(mem, r, offset, &mut bytes);
+    core::primitive::f64::from_le_bytes(bytes).into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_blob<M: Memory>(mem: &mut M, r: Value, offset: u64, len: u32) -> Value {
+    let blob_val = crate::memory::alloc_blob(mem, crate::types::Bytes(len));
+    let blob = blob_val.as_blob_mut();
+    let bytes: &mut [u8] = core::slice::from_raw_parts_mut(blob.payload_addr(), len as usize);
+    region_load(mem, r, offset, bytes);
+    blob_val
```

Comment:
Required allocation barrier:

```suggestion
    allocation_barrier(blob_val);
    blob_val
```